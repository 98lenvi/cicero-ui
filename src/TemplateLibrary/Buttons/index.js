/* React */
import React from 'react';
import PropTypes from 'prop-types';

/* Styling */
import styled from 'styled-components';
import { Button } from 'semantic-ui-react';

const UploadButton = styled.button`
  position: relative;
  font-weight: 300;
  text-align: right;
  text-decoration: none;
  font-size: 14px;
  color: #76777D;
  border: 0;
  cursor: pointer;
  &:hover {
    color: #0361DE;
    text-decoration: underline;
  }
  &:focus {
    outline: none;
    color: #3089FF;
    text-decoration: underline;
  }
  &:active {
    color: #3089FF;
    text-decoration: underline;
  }
`;

const ImportButton = styled(UploadButton)`
  margin-bottom: 5px;
`;

const AddClauseBtn = styled(Button)`
  margin: 5px auto 0 !important;
  width: 96% !important;
`;

export const ImportComponent = props => (
    <ImportButton
        onClick={props.importInput}
        href="javascript:void(0);"
        className="importButton"
    >
        Import from VS Code
    </ImportButton>);

export const UploadComponent = props => (
    <UploadButton
        onClick={props.uploadInput}
        href="javascript:void(0);"
        className="uploadButton"
    >
        Upload CTA file
    </UploadButton>);

export const NewClauseComponent = props => (
    <AddClauseBtn
        content="New Clause Template"
        color="blue"
        fluid
        icon="plus"
        id="addClauseBtn"
        onClick={props.addTempInput}
        className="addTemplateButton"
    />);

ImportComponent.propTypes = { importInput: PropTypes.func };
UploadComponent.propTypes = { uploadInput: PropTypes.func };
NewClauseComponent.propTypes = { addTempInput: PropTypes.func };
