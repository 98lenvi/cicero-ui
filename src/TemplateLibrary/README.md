## Cicero-UI - TemplateLibrary

### Usage

```
npm install @accordproject/cicero-ui
```

```
import { TemplateLibrary } from '@accordproject/cicero-ui';

const LibraryComponent = props => (
    <TemplateLibrary
        templates={props.templatesArray}
        upload={props.uploadCTA}
        import={props.importTemplate}
        addTemp={props.addNewTemplate}
        addToCont={props.addToContract}
    />
);
```

### Props

- `templates` : An `array` which contains template objects with the following keys: `uir`, `name`, `version`, `description`.
- `upload` : A function which calls for upload functionality within the app this component is embedded in.
- `import` : A function which calls for import functionality within the app this component is embedded in.
- `addTemp` : A function which adds a new blank template to the array of templates in the Redux store of the app this component is embedded in.
- `addToCont` : A function which calls for adding the selected template to the Redux store in the app this component is embedded in. This will result in another component having use of its data.

### Specifications

This component is built to have the following dimensions:

```
height: 700px;
width: 485px;
```