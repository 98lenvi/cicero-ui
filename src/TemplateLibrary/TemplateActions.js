import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card, Icon } from 'semantic-ui-react';

const ActionsContainer = styled(Card.Content)`
  padding: 0 !important;
  background-color: #F9F9F9 !important;
  max-height: 30px;
`;

const TemplateBtn = styled.a`
  padding: 5px 10px;
  display: inline-block;
  color: #484848;
  font-family: "IBM Plex Sans";
  font-size: 12px;
  font-weight: bold;
`;

const AddToContractBtn = styled(TemplateBtn)`
  width: 60%;
  border-right: 1px solid #E1E5EB;
  cursor: pointer;
  &:hover {
    color: #3087CB;
  }
`;

const DetailsBtn = styled(TemplateBtn)`
  float: right;
  width: 40%;
  font-size: 12px;
  font-weight: 300;
  text-align: center;
`;

/**
 * A Template Actions component that will provide each template
 * with functionality.
 */
class TemplateActions extends React.Component {
  /**
     * Render this React component
     * @return {*} the react component
  */
  render() {
    return (
        <ActionsContainer>
        <div>
          <AddToContractBtn className="adToContractButton" onClick={() => this.props.addToCont(this.props.uriKey)} >
            <Icon name="plus" />
            Add to contract
          </AddToContractBtn>
          <DetailsBtn onClick={() => this.props.handleViewDetails(this.props.uriKey)}>
            Details
          </DetailsBtn>
        </div>
      </ActionsContainer>
    );
  }
}

/**
 * The property types for this component
 */
TemplateActions.propTypes = {
  addToCont: PropTypes.func,
  handleViewDetails: PropTypes.func,
  uriKey: PropTypes.string,
};

export default TemplateActions;
