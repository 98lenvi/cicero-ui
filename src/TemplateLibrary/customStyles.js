import styled from 'styled-components';

export const CustomStylesWrapper = styled.div`
.templateListTitle{
};
.templateSearchInput{
};
.templateCardsDiv{
};
.templateCard{
};
.templateCardTitle{
};
.templateCardDesc{
};
.templateActions{
};
.templateAddToContractButton{
};
.templateDetailsButton{
};
.templateImportButton{
};
.templateUploadButton{
};
.addTemplateButton{
};
`