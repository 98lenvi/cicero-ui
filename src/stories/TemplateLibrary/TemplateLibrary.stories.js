import React from "react";
import docs from "../../TemplateLibrary/README.md";
import { withA11y } from "@storybook/addon-a11y";
import { withKnobs, text } from "@storybook/addon-knobs";
import TemplateLibrary from "../../TemplateLibrary";

export default { title: "'Components/Template Library" };
export const templateLibrary = () => {
  const libraryPropsObject = {
    ACTION_BUTTON: text("ACTION_BUTTON", ""),
    ACTION_BUTTON_BG: text("ACTION_BUTTON_BG", ""),
    ACTION_BUTTON_BORDER: text("ACTION_BUTTON_BORDER", ""),
    HEADER_TITLE: text("HEADER_TITLE", ""),
    SEARCH_COLOR: text("SEARCH_COLOR", ""),
    TEMPLATE_BACKGROUND: text("TEMPLATE_BACKGROUND", ""),
    TEMPLATE_DESCRIPTION: text("TEMPLATE_DESCRIPTION", ""),
    TEMPLATE_TITLE: text("TEMPLATE_TITLE", "")
  };
  const templates = [
    {
      uri: text(
        "First template's Uri",
        "https://templates.accordproject.org/archives/acceptance-of-delivery@0.13.1.cta"
      ),
      name: text("First template's Name", "Acceptance of Delivery"),
      version: text("First template's Version", "^0.20.10"),
      description: text(
        "First template's Description",
        "This clause allows the receiver of goods to inspect them for a given time period after delivery."
      )
    },
    {
      uri:
        "https://templates.accordproject.org/archives/car-rental-tr@0.10.1.cta",
      name: "Car Rental (TR)",
      version: "^0.20.10",
      description: "Ta Simple Car Rental Contract in Turkish Language"
    },
    {
      uri:
        "https://templates.accordproject.org/archives/certificate-of-incorporation@0.3.1.cta",
      name: "Certificate Of Incorporation",
      version: "^0.20.10",
      description: "This is a sample Certificate of Incorporation."
    }
  ];
  const mockImport = () => {
    alert(
      "A function which calls for import functionality within the app this component is embedded in"
    );
  };
  const mockUpload = () => {
    alert(
      "A function which calls for upload functionality within the app this component is embedded in"
    );
  };
  const mockAddTemp = () => {
    alert(
      "A function which adds a new blank template to the array of templates in the Redux store of the app this component is embedded in."
    );
  };
  const mockAddToCont = () => {
    alert(
      " A function which calls for adding the selected template to the Redux store in the app this component is embedded in. This will result in another component having use of its data."
    );
  };
  return (
    <TemplateLibrary
      templates={templates}
      upload={mockUpload}
      import={mockImport}
      addTemp={mockAddTemp}
      addToCont={mockAddToCont}
      libraryProps={libraryPropsObject}
    />
  );
};

templateLibrary.story = {
  component: templateLibrary,
  decorators: [withA11y, withKnobs],
  parameters: {
    notes: docs
  }
};
