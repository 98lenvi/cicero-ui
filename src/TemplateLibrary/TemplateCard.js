import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Card, Image } from 'semantic-ui-react';

import TemplateActions from './TemplateActions';

const CardContainer = styled(Card)`
  margin: 10px 0 !important;
  background-color: ${props => props.templateBackground || null} !important;
  position: relative;
  text-align: left;
  min-height: 120px;
  font-size: 14px !important;
  box-shadow: 0 1px 9px 0 rgba(0,0,0,0.1) !important;
`;

const Title = styled.div`
  height: 16px;
  color: ${props => props.templateTitle || null};
  font-family: Graphik;
  font-size: 16px;
  font-weight: 600;
  line-height: 16px;
  margin-bottom: 5px;
`;

const TemplateLogo = styled(Image)`
  position: absolute !important;
  top: 13px;
  right: 16px;
  max-height: 23px;
`;

const Version = styled.span`
  margin-left: 10px;
  font-size: 12px;
  font-weight: 300;
`;

const DescriptionContainer = styled(Card.Description)`
  max-width: 400px;
  margin: auto;
  font-size: 0.9em;
  color: ${props => props.templateDescription || null} !important;
`;

/**
 * A Template Card component that will display the each template
 * and it's details.
 */
class TemplateCard extends React.Component {
  /**
     * Render this React component
     * @return {*} the react component
  */
  render() {
    const { template } = this.props;
    return (
        <CardContainer fluid key={template.uri} templateBackground={this.props.templateBackground}>
            <Card.Content>
              <TemplateLogo src={template.icon} />
              <Title templateTitle={this.props.templateTitle}>
                {template.name}
                <Version>v {template.version}</Version>
              </Title>
              <DescriptionContainer templateDescription={this.props.templateDescription}>
                {template.description}
              </DescriptionContainer>
            </Card.Content>
            <TemplateActions
              actionBtnBkgrd={this.props.actionBtnBkgrd}
              actionBtnColor={this.props.actionBtnColor}
              actionBtnBorder={this.props.actionBtnBorder}
              addToCont={this.props.addToCont}
              uriKey={template.uri}
              handleViewDetails={this.props.handleViewTemplate}
              className="templateAction"
            />
        </CardContainer>
    );
  }
}

/**
 * The property types for this component
 */
TemplateCard.propTypes = {
  templateDescription: PropTypes.string,
  template: PropTypes.object,
  addToCont: PropTypes.func,
  handleViewTemplate: PropTypes.func,
  actionBtnColor: PropTypes.string,
  actionBtnBkgrd: PropTypes.string,
  actionBtnBorder: PropTypes.string,
  templateBackground: PropTypes.string,
  templateTitle: PropTypes.string,
};

export default TemplateCard;
