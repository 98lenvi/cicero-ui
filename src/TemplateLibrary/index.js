import _ from 'lodash';

/* React */
import React from 'react';
import PropTypes from 'prop-types';

/* Styling */
import styled from 'styled-components';
import { Card, Input } from 'semantic-ui-react';

/* Internal */
import TemplateCard from './TemplateCard';
import { ImportComponent, UploadComponent, NewClauseComponent } from './Buttons';

const TemplatesWrapper = styled.div`
  font-family: 'IBM Plex Sans', sans-serif;
  position: relative;
  max-width: 355px;

  display: grid;
  grid-template-areas:  "header"
                        "functionTemps";
  grid-template-rows:    55px auto;
  grid-template-columns: 100%;
`;

const Header = styled.div`
  position: relative;
  font-family: 'IBM Plex Sans', sans-serif;
  font-weight: 800;
  font-size: 16px;
  margin-bottom: 10px;
  
  display: grid;
  grid-area: header;
  grid-template-columns: 1fr 1fr;
  grid-template-areas: "title imports";
`;

const HeaderTitle = styled.p`
  text-align: left;
`;

const HeaderImports = styled.div`
  display: flex;
  flex-direction: column;
`;

const Functionality = styled.div`
  font-family: 'IBM Plex Sans', sans-serif;
  max-width: 430px;
  display: flex;
  flex-direction: column;
`;

const SearchInput = styled(Input)`
  margin: 5px 0 !important;
  width: 100% !important;
`;

const TemplateCards = styled(Card.Group)`
  margin: 0 !important;
  width: 100%;
`;

/**
 * A Template Library component that will display the filtered list of templates
 * and provide drag-and-drop functionality.
 */

class TemplateLibraryComponent extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      query: '',
    };
    this.onQueryChange = this.onQueryChange.bind(this);
  }

  static propTypes = {
    upload: PropTypes.func,
    import: PropTypes.func,
    addTemp: PropTypes.func,
    addToCont: PropTypes.func,
    handleViewTemplate: PropTypes.func,
    templates: PropTypes.arrayOf(PropTypes.object),
  }

  onQueryChange(e, input) {
    const query = input.value.toLowerCase().trim();
    if (query !== this.state.query) {
      this.setState({ query });
    }
  }

  filterTemplates(templates) {
    const { query } = this.state;
    let filteredTemplates = templates;
    if (query.length) {
      const regex = new RegExp(query, 'i');
      filteredTemplates = _.filter(filteredTemplates, t => (
        t.name.match(regex) || t.uri.match(regex)
      ));
    }
    return filteredTemplates;
  }

  /**
   * Render this React component
   * @return {*} the react component
   */
  render() {
    const filtered = this.filterTemplates(this.props.templates);
    return (
      <TemplatesWrapper>
        <Header>
          <HeaderTitle>Clause Templates</HeaderTitle>
          <HeaderImports>
            {this.props.import
            && <ImportComponent importInput={this.props.import} />}
            {this.props.upload
            && <UploadComponent uploadInput={this.props.upload} />}
          </HeaderImports>
        </Header>
        <Functionality>
          <SearchInput className="icon" fluid icon="search" placeholder="Search..." onChange={this.onQueryChange} />
          {this.props.addTemp
          && <NewClauseComponent addTempInput={this.props.addTemp} />}
        </Functionality>
        <TemplateCards>
          {_.sortBy(filtered, ['name']).map(t => (
            <TemplateCard
              key={t.uri}
              addToCont={this.props.addToCont}
              template={t}
              handleViewTemplate={this.props.handleViewTemplate}
              className="templateCard"
            />
          ))}
        </TemplateCards>
      </TemplatesWrapper>
    );
  }
}

export default TemplateLibraryComponent;
