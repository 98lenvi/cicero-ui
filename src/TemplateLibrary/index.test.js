import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

import TemplateLibrary from './index';

Enzyme.configure({ adapter: new Adapter() });

describe('<TemplateLibrary />', () => {
  describe('on initialization', () => {
    it('renders page correctly', () => {
      const component = shallow(<TemplateLibrary />);
      const tree = toJson(component);
      expect(tree).toMatchSnapshot();
    });
  });
});
