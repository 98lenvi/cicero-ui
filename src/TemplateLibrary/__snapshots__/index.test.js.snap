// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateLibrary /> on initialization renders page correctly 1`] = `
<div>
  <styled.div>
    <styled.div>
      Clause Templates
    </styled.div>
    <styled.div>
      <Styled(Input)
        className="icon"
        fluid={true}
        icon="search"
        onChange={[Function]}
        placeholder="Search..."
      />
      <Styled(Button)
        color="blue"
        content="New Clause Template"
        fluid={true}
        icon="plus"
        id="addClauseBtn"
      />
    </styled.div>
    <Styled(CardGroup) />
  </styled.div>
</div>
`;
